src/GafferOSL/ShadingEngine.cpp

```diff
@@ -632,10 +597,70 @@ class ShadingResults
 			}
 		};
 
+		DebugResult findDebugResult(const DebugParameters *parameters)
+		{
+			ResultCache::const_accessor access;
+
+			if( m_resultCache.find( access, parameters->name ) )
+			{
+				return access->second;
+			}
+			else
+			{
+				ResultCache::accessor insertAccessor;
+				m_resultCache.insert( insertAccessor, parameters->name );
+
+				vector<DebugResult>::iterator it = lower_bound(
+					m_debugResults.begin(),
+					m_debugResults.end(),
+					parameters->name
+				);
+
+				if( it == m_debugResults.end() || it->name != parameters->name )
+				{
+					DebugResult result;
+					result.name = parameters->name;
+					result.type = parameters->type != ustring() ? TypeDesc( parameters->type.c_str() ) : TypeDesc::TypeColor;
+					result.type.arraylen = m_ci->size();
+					DataPtr data = dataFromTypeDesc( result.type, result.basePointer );
+					if( !data )
+					{
+						throw IECore::Exception( "Unsupported type specified in debug() closure." );
+					}
+					result.type.unarray(); // so we can use convert_value
+					m_results->writable()[result.name.c_str()] = data;
```

Comment:
As far as I can tell, this is unprotected from concurrent access, as is the mutation of `m_debugResults` on the line below. Am I missing something?

Reply:
You're right this isn't safe. My faulty assumption was the scope of the concurrent_hash_map accessor would provide sufficient locking. It needs to be a global lock on m_results!